File: deepseek_infra/infra/native_runtime/mutation_request.py
"""Fail-closed control-mutation-request-v1 verifier.

This module is the Python oracle for the frozen native-runtime v17 corpus.
It verifies canonical JSON mutation proposals. It does not apply production
mutations, install a live epoch, or hold Federation root private keys.
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Mapping

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey, Ed25519PublicKey
from cryptography.exceptions import InvalidSignature

from deepseek_infra.infra.native_runtime.authority_request import (
    ALLOWED_DOMAINS,
    SIGNATURE_ALGORITHM,
    AuthorityRequestError,
    _CONTROL_ID_PATTERN,
    _FLEET_ID_PATTERN,
    _HEX64_PATTERN,
    _SIGNER_KEY_ID_PATTERN,
    _b64url_decode,
    _b64url_encode,
    _decode_single_json,
    _parse_timestamp,
    _reject_secrets,
    canonical_authority_request_bytes,
    signer_key_id_for_public_key,
)

MUTATION_REQUEST_SCHEMA = "control-mutation-request-v1"
MUTATION_REQUEST_SCHEMA_VERSION = 1
SIGNATURE_DOMAIN = b"deepseek-infra:control-mutation-request-v1\x00"
MAX_MUTATION_REQUEST_BYTES = 16 * 1024
MAX_LIFETIME_SECONDS = 300
ALLOWED_OPERATIONS = frozenset({"propose-mutation"})
ALLOWED_INTENTS = frozenset({"shadow-compare"})
PAYLOAD_FIELDS = ("intent", "recordId", "revision", "state")
MUTATION_REQUEST_FIELDS = (
    "actionId",
    "digest",
    "domain",
    "environment",
    "executionEpoch",
    "expiresAt",
    "fencingToken",
    "fleetId",
    "issuedAt",
    "mode",
    "nonce",
    "operation",
    "operationId",
    "payload",
    "payloadDigest",
    "requestId",
    "revision",
    "role",
    "runtime",
    "schema",
    "schemaVersion",
    "signature",
    "signatureAlgorithm",
    "signerKeyId",
)


class MutationRequestError(AuthorityRequestError):
    pass


@dataclass(frozen=True)
class MutationRequestContext:
    now: datetime
    signer_public_key: str
    signer_key_id: str
    expected_domain: str
    expected_operation: str
    expected_runtime: str
    expected_mode: str
    expected_fleet_id: str
    expected_environment: str
    expected_role: str
    current_fencing_token: int
    live_epoch: int
    seen_request_ids: frozenset[str]
    seen_nonces: frozenset[str]
    seen_operation_digests: Mapping[str, str]
    max_future_skew_seconds: int = 30


def mutation_request_digest(value: Mapping[str, Any]) -> str:
    unsigned = {key: item for key, item in value.items() if key not in {"signature", "digest"}}
    return "sha256:" + hashlib.sha256(canonical_authority_request_bytes(unsigned)).hexdigest()


def sign_mutation_request(
    unsigned: Mapping[str, Any],
    *,
    private_key: Ed25519PrivateKey,
    public_key: str,
) -> dict[str, Any]:
    payload = dict(unsigned)
    if "signature" in payload:
        raise MutationRequestError("MUTATION_REQUEST_INVALID")
    payload["signerKeyId"] = signer_key_id_for_public_key(public_key)
    payload["signatureAlgorithm"] = SIGNATURE_ALGORITHM
    payload["payloadDigest"] = _payload_digest(payload.get("payload"))
    payload["digest"] = mutation_request_digest(payload)
    message = SIGNATURE_DOMAIN + canonical_authority_request_bytes(payload)
    signed = dict(payload)
    signed["signature"] = _b64url_encode(private_key.sign(message))
    return signed


def verify_mutation_request_document(raw: bytes, context: MutationRequestContext) -> dict[str, Any]:
    if not raw:
        raise MutationRequestError("MUTATION_REQUEST_INVALID")
    if len(raw) > MAX_MUTATION_REQUEST_BYTES:
        raise MutationRequestError("MUTATION_REQUEST_TOO_LARGE")
    try:
        document = _decode_single_json(raw)
        if type(document) is not dict:
            raise MutationRequestError("MUTATION_REQUEST_INVALID")
        canonical = canonical_authority_request_bytes(document)
    except AuthorityRequestError as exc:
        raise MutationRequestError("MUTATION_REQUEST_INVALID") from exc
    if canonical != raw:
        raise MutationRequestError("MUTATION_REQUEST_CANONICAL_MISMATCH")
    if tuple(sorted(document)) != MUTATION_REQUEST_FIELDS:
        raise MutationRequestError("MUTATION_REQUEST_FIELDS_INVALID")
    try:
        _reject_secrets(document)
    except AuthorityRequestError as exc:
        if exc.code == "AUTHORITY_REQUEST_SECRET_DETECTED":
            raise MutationRequestError("MUTATION_REQUEST_SECRET_DETECTED") from exc
        raise MutationRequestError("MUTATION_REQUEST_INVALID") from exc
    _verify_envelope(document, context)
    _verify_signature(document, context)
    return document


def _payload_digest(payload: Any) -> str:
    return "sha256:" + hashlib.sha256(canonical_authority_request_bytes(payload)).hexdigest()


def _verify_envelope(document: Mapping[str, Any], context: MutationRequestContext) -> None:
    if document["schema"] != MUTATION_REQUEST_SCHEMA or document["schemaVersion"] != MUTATION_REQUEST_SCHEMA_VERSION:
        raise MutationRequestError("MUTATION_REQUEST_SCHEMA_INVALID")
    if document["operation"] not in ALLOWED_OPERATIONS or document["operation"] != context.expected_operation:
        raise MutationRequestError("MUTATION_REQUEST_OPERATION_INVALID")
    if document["domain"] not in ALLOWED_DOMAINS or document["domain"] != context.expected_domain:
        raise MutationRequestError("MUTATION_REQUEST_DOMAIN_MISMATCH")
    if document["runtime"] != "go" or document["runtime"] != context.expected_runtime:
        raise MutationRequestError("MUTATION_REQUEST_RUNTIME_MISMATCH")
    if document["mode"] != "shadow" or document["mode"] != context.expected_mode:
        raise MutationRequestError("MUTATION_REQUEST_MODE_MISMATCH")
    if not _FLEET_ID_PATTERN.fullmatch(str(document["fleetId"])) or document["fleetId"] != context.expected_fleet_id:
        raise MutationRequestError("MUTATION_REQUEST_FLEET_MISMATCH")
    if type(document["environment"]) is not str or not document["environment"] or document["environment"] != context.expected_environment:
        raise MutationRequestError("MUTATION_REQUEST_ENVIRONMENT_MISMATCH")
    if document["role"] != "control-plane" or document["role"] != context.expected_role:
        raise MutationRequestError("MUTATION_REQUEST_ROLE_MISMATCH")
    if type(document["actionId"]) is not str or not _CONTROL_ID_PATTERN.fullmatch(document["actionId"]):
        raise MutationRequestError("EMPTY_ACTION_ID")
    if type(document["executionEpoch"]) is not int or isinstance(document["executionEpoch"], bool) or document["executionEpoch"] < 1:
        raise MutationRequestError("ZERO_EXECUTION_EPOCH")
    if type(document["revision"]) is not int or isinstance(document["revision"], bool) or document["revision"] < 1:
        raise MutationRequestError("MUTATION_REQUEST_INVALID")
    if type(document["fencingToken"]) is not int or isinstance(document["fencingToken"], bool) or document["fencingToken"] < 1:
        raise MutationRequestError("MUTATION_REQUEST_STALE_FENCING_TOKEN")
    if document["fencingToken"] != context.current_fencing_token:
        raise MutationRequestError("MUTATION_REQUEST_STALE_FENCING_TOKEN")
    if document["executionEpoch"] <= context.live_epoch:
        raise MutationRequestError("STALE_EXECUTION_EPOCH")
    if type(document["requestId"]) is not str or not _HEX64_PATTERN.fullmatch(document["requestId"]):
        raise MutationRequestError("MUTATION_REQUEST_INVALID")
    if type(document["nonce"]) is not str or not _HEX64_PATTERN.fullmatch(document["nonce"]):
        raise MutationRequestError("MUTATION_REQUEST_INVALID")
    if type(document["operationId"]) is not str or not _HEX64_PATTERN.fullmatch(document["operationId"]):
        raise MutationRequestError("MUTATION_REQUEST_INVALID")
    if document["requestId"] in context.seen_request_ids:
        raise MutationRequestError("MUTATION_REQUEST_REPLAY")
    if document["nonce"] in context.seen_nonces:
        raise MutationRequestError("MUTATION_REQUEST_NONCE_REUSE")
    payload = document["payload"]
    if type(payload) is not dict or tuple(sorted(payload)) != PAYLOAD_FIELDS:
        raise MutationRequestError("MUTATION_REQUEST_INVALID")
    if payload["intent"] not in ALLOWED_INTENTS:
        raise MutationRequestError("MUTATION_REQUEST_INVALID")
    if type(payload["recordId"]) is not str or not _CONTROL_ID_PATTERN.fullmatch(payload["recordId"]):
        raise MutationRequestError("MUTATION_REQUEST_INVALID")
    if type(payload["revision"]) is not int or isinstance(payload["revision"], bool) or payload["revision"] < 1:
        raise MutationRequestError("MUTATION_REQUEST_INVALID")
    if type(payload["state"]) is not str or not payload["state"]:
        raise MutationRequestError("MUTATION_REQUEST_INVALID")
    expected_payload_digest = _payload_digest(payload)
    if document["payloadDigest"] != expected_payload_digest:
        raise MutationRequestError("MUTATION_REQUEST_PAYLOAD_DIGEST_MISMATCH")
    seen_digest = context.seen_operation_digests.get(document["operationId"])
    if seen_digest is not None and seen_digest != expected_payload_digest:
        raise MutationRequestError("MUTATION_REQUEST_REPLAY_CONFLICT")
    if document["digest"] != mutation_request_digest(document):
        raise MutationRequestError("MUTATION_REQUEST_DIGEST_MISMATCH")
    try:
        issued_at = _parse_timestamp(document["issuedAt"])
        expires_at = _parse_timestamp(document["expiresAt"])
    except AuthorityRequestError as exc:
        raise MutationRequestError("MUTATION_REQUEST_INVALID") from exc
    now = context.now.astimezone(timezone.utc).replace(microsecond=0)
    if expires_at <= issued_at or (expires_at - issued_at).total_seconds() > MAX_LIFETIME_SECONDS:
        raise MutationRequestError("MUTATION_REQUEST_INVALID")
    if expires_at <= now:
        raise MutationRequestError("MUTATION_REQUEST_EXPIRED")
    skew = max(0, int(context.max_future_skew_seconds))
    if (issued_at - now).total_seconds() > skew:
        raise MutationRequestError("MUTATION_REQUEST_FUTURE_SKEW")


def _verify_signature(document: Mapping[str, Any], context: MutationRequestContext) -> None:
    if document["signatureAlgorithm"] != SIGNATURE_ALGORITHM:
        raise MutationRequestError("MUTATION_REQUEST_SIGNATURE_INVALID")
    if document["signerKeyId"] != context.signer_key_id or not _SIGNER_KEY_ID_PATTERN.fullmatch(document["signerKeyId"]):
        raise MutationRequestError("MUTATION_REQUEST_SIGNER_MISMATCH")
    signature = _b64url_decode(document["signature"], expected_length=64)
    public = _b64url_decode(context.signer_public_key, expected_length=32)
    if signature is None or public is None:
        raise MutationRequestError("MUTATION_REQUEST_SIGNATURE_INVALID")
    unsigned = {key: value for key, value in document.items() if key != "signature"}
    message = SIGNATURE_DOMAIN + canonical_authority_request_bytes(unsigned)
    try:
        Ed25519PublicKey.from_public_bytes(public).verify(signature, message)
    except InvalidSignature as exc:
        raise MutationRequestError("MUTATION_REQUEST_SIGNATURE_INVALID") from exc

File: rust/crates/deepseek-worker/src/mutation_request.rs
use crate::authority_request::{AuthorityRequestError, canonical_json_bytes};
use base64::Engine as _;
use base64::engine::general_purpose::URL_SAFE_NO_PAD;
use ed25519_dalek::{Signature, VerifyingKey};
use serde_json::{Map, Value};
use sha2::{Digest, Sha256};
use std::collections::{BTreeSet, HashMap, HashSet};

pub const MUTATION_REQUEST_SCHEMA: &str = "control-mutation-request-v1";
pub const MAX_MUTATION_REQUEST_BYTES: usize = 16 * 1024;
const SIGNATURE_DOMAIN: &[u8] = b"deepseek-infra:control-mutation-request-v1\x00";
const MAX_LIFETIME_SECONDS: i64 = 300;
const MUTATION_REQUEST_FIELDS: &[&str] = &[
    "actionId",
    "digest",
    "domain",
    "environment",
    "executionEpoch",
    "expiresAt",
    "fencingToken",
    "fleetId",
    "issuedAt",
    "mode",
    "nonce",
    "operation",
    "operationId",
    "payload",
    "payloadDigest",
    "requestId",
    "revision",
    "role",
    "runtime",
    "schema",
    "schemaVersion",
    "signature",
    "signatureAlgorithm",
    "signerKeyId",
];
const PAYLOAD_FIELDS: &[&str] = &["intent", "recordId", "revision", "state"];

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MutationRequestError {
    pub code: &'static str,
}

impl MutationRequestError {
    fn new(code: &'static str) -> Self {
        Self { code }
    }
}

impl From<AuthorityRequestError> for MutationRequestError {
    fn from(error: AuthorityRequestError) -> Self {
        Self { code: error.code }
    }
}

#[derive(Debug, Clone)]
pub struct MutationRequestContext<'a> {
    pub now: &'a str,
    pub signer_public_key: &'a str,
    pub signer_key_id: &'a str,
    pub expected_domain: &'a str,
    pub expected_operation: &'a str,
    pub expected_runtime: &'a str,
    pub expected_mode: &'a str,
    pub expected_fleet_id: &'a str,
    pub expected_environment: &'a str,
    pub expected_role: &'a str,
    pub current_fencing_token: i64,
    pub live_epoch: i64,
    pub seen_request_ids: HashSet<String>,
    pub seen_nonces: HashSet<String>,
    pub seen_operation_digests: HashMap<String, String>,
    pub max_future_skew_seconds: i64,
}

pub fn verify_mutation_request_document(
    raw: &[u8],
    context: &MutationRequestContext<'_>,
) -> Result<Value, MutationRequestError> {
    if raw.is_empty() {
        return Err(MutationRequestError::new("MUTATION_REQUEST_INVALID"));
    }
    if raw.len() > MAX_MUTATION_REQUEST_BYTES {
        return Err(MutationRequestError::new("MUTATION_REQUEST_TOO_LARGE"));
    }
    let document: Value = serde_json::from_slice(raw)
        .map_err(|_| MutationRequestError::new("MUTATION_REQUEST_INVALID"))?;
    let canonical = canonical_json_bytes(&document)
        .map_err(|_| MutationRequestError::new("MUTATION_REQUEST_INVALID"))?;
    if canonical != raw {
        return Err(MutationRequestError::new(
            "MUTATION_REQUEST_CANONICAL_MISMATCH",
        ));
    }
    let object = document
        .as_object()
        .ok_or_else(|| MutationRequestError::new("MUTATION_REQUEST_INVALID"))?;
    let keys: BTreeSet<&str> = object.keys().map(String::as_str).collect();
    if keys.len() != MUTATION_REQUEST_FIELDS.len()
        || MUTATION_REQUEST_FIELDS
            .iter()
            .any(|field| !keys.contains(field))
    {
        return Err(MutationRequestError::new("MUTATION_REQUEST_FIELDS_INVALID"));
    }
    reject_secrets(&document)?;
    verify_envelope(object, context)?;
    verify_signature(object, context)?;
    Ok(document)
}

fn verify_envelope(
    document: &Map<String, Value>,
    context: &MutationRequestContext<'_>,
) -> Result<(), MutationRequestError> {
    if string_field(document, "schema") != Some(MUTATION_REQUEST_SCHEMA)
        || document.get("schemaVersion") != Some(&Value::from(1))
    {
        return Err(MutationRequestError::new("MUTATION_REQUEST_SCHEMA_INVALID"));
    }
    if string_field(document, "operation") != Some(context.expected_operation)
        || string_field(document, "operation") != Some("propose-mutation")
    {
        return Err(MutationRequestError::new(
            "MUTATION_REQUEST_OPERATION_INVALID",
        ));
    }
    if string_field(document, "domain") != Some(context.expected_domain) {
        return Err(MutationRequestError::new(
            "MUTATION_REQUEST_DOMAIN_MISMATCH",
        ));
    }
    if string_field(document, "runtime") != Some("go")
        || string_field(document, "runtime") != Some(context.expected_runtime)
    {
        return Err(MutationRequestError::new(
            "MUTATION_REQUEST_RUNTIME_MISMATCH",
        ));
    }
    if string_field(document, "mode") != Some("shadow")
        || string_field(document, "mode") != Some(context.expected_mode)
    {
        return Err(MutationRequestError::new("MUTATION_REQUEST_MODE_MISMATCH"));
    }
    let fleet_id = string_field(document, "fleetId").unwrap_or("");
    if !valid_fleet_id(fleet_id) || fleet_id != context.expected_fleet_id {
        return Err(MutationRequestError::new("MUTATION_REQUEST_FLEET_MISMATCH"));
    }
    if string_field(document, "environment") != Some(context.expected_environment)
        || context.expected_environment.is_empty()
    {
        return Err(MutationRequestError::new(
            "MUTATION_REQUEST_ENVIRONMENT_MISMATCH",
        ));
    }
    if string_field(document, "role") != Some("control-plane")
        || string_field(document, "role") != Some(context.expected_role)
    {
        return Err(MutationRequestError::new("MUTATION_REQUEST_ROLE_MISMATCH"));
    }
    let action_id = string_field(document, "actionId").unwrap_or("");
    if !valid_control_id(action_id) {
        return Err(MutationRequestError::new("EMPTY_ACTION_ID"));
    }
    let epoch = integer_field(document, "executionEpoch")?;
    if epoch < 1 {
        return Err(MutationRequestError::new("ZERO_EXECUTION_EPOCH"));
    }
    if integer_field(document, "revision")? < 1 {
        return Err(MutationRequestError::new("MUTATION_REQUEST_INVALID"));
    }
    let fencing_token = integer_field(document, "fencingToken")?;
    if fencing_token < 1 || fencing_token != context.current_fencing_token {
        return Err(MutationRequestError::new(
            "MUTATION_REQUEST_STALE_FENCING_TOKEN",
        ));
    }
    if epoch <= context.live_epoch {
        return Err(MutationRequestError::new("STALE_EXECUTION_EPOCH"));
    }
    let request_id = string_field(document, "requestId").unwrap_or("");
    let nonce = string_field(document, "nonce").unwrap_or("");
    let operation_id = string_field(document, "operationId").unwrap_or("");
    if !valid_hex64(request_id) || !valid_hex64(nonce) || !valid_hex64(operation_id) {
        return Err(MutationRequestError::new("MUTATION_REQUEST_INVALID"));
    }
    if context.seen_request_ids.contains(request_id) {
        return Err(MutationRequestError::new("MUTATION_REQUEST_REPLAY"));
    }
    if context.seen_nonces.contains(nonce) {
        return Err(MutationRequestError::new("MUTATION_REQUEST_NONCE_REUSE"));
    }
    let payload = document
        .get("payload")
        .and_then(Value::as_object)
        .ok_or_else(|| MutationRequestError::new("MUTATION_REQUEST_INVALID"))?;
    let payload_keys: BTreeSet<&str> = payload.keys().map(String::as_str).collect();
    if payload_keys.len() != PAYLOAD_FIELDS.len()
        || PAYLOAD_FIELDS
            .iter()
            .any(|field| !payload_keys.contains(field))
    {
        return Err(MutationRequestError::new("MUTATION_REQUEST_INVALID"));
    }
    if string_field(payload, "intent") != Some("shadow-compare") {
        return Err(MutationRequestError::new("MUTATION_REQUEST_INVALID"));
    }
    if !valid_control_id(string_field(payload, "recordId").unwrap_or("")) {
        return Err(MutationRequestError::new("MUTATION_REQUEST_INVALID"));
    }
    if integer_field(payload, "revision")? < 1 {
        return Err(MutationRequestError::new("MUTATION_REQUEST_INVALID"));
    }
    if string_field(payload, "state").unwrap_or("").is_empty() {
        return Err(MutationRequestError::new("MUTATION_REQUEST_INVALID"));
    }
    let payload_digest = typed_digest(&Value::Object(payload.clone()))?;
    if string_field(document, "payloadDigest") != Some(payload_digest.as_str()) {
        return Err(MutationRequestError::new(
            "MUTATION_REQUEST_PAYLOAD_DIGEST_MISMATCH",
        ));
    }
    if let Some(seen) = context.seen_operation_digests.get(operation_id) {
        if seen != payload_digest.as_str() {
            return Err(MutationRequestError::new(
                "MUTATION_REQUEST_REPLAY_CONFLICT",
            ));
        }
    }
    let digest = mutation_digest(document)?;
    if string_field(document, "digest") != Some(digest.as_str()) {
        return Err(MutationRequestError::new(
            "MUTATION_REQUEST_DIGEST_MISMATCH",
        ));
    }
    let issued_at = parse_utc_z(string_field(document, "issuedAt").unwrap_or(""))?;
    let expires_at = parse_utc_z(string_field(document, "expiresAt").unwrap_or(""))?;
    let now = parse_utc_z(context.now)?;
    if expires_at <= issued_at || expires_at - issued_at > MAX_LIFETIME_SECONDS {
        return Err(MutationRequestError::new("MUTATION_REQUEST_INVALID"));
    }
    if expires_at <= now {
        return Err(MutationRequestError::new("MUTATION_REQUEST_EXPIRED"));
    }
    let skew = context.max_future_skew_seconds.max(0);
    if issued_at.saturating_sub(now) > skew {
        return Err(MutationRequestError::new("MUTATION_REQUEST_FUTURE_SKEW"));
    }
    Ok(())
}

fn verify_signature(
    document: &Map<String, Value>,
    context: &MutationRequestContext<'_>,
) -> Result<(), MutationRequestError> {
    if string_field(document, "signatureAlgorithm") != Some("Ed25519") {
        return Err(MutationRequestError::new(
            "MUTATION_REQUEST_SIGNATURE_INVALID",
        ));
    }
    let signer_key_id = string_field(document, "signerKeyId").unwrap_or("");
    if signer_key_id != context.signer_key_id || !valid_signer_key_id(signer_key_id) {
        return Err(MutationRequestError::new(
            "MUTATION_REQUEST_SIGNER_MISMATCH",
        ));
    }
    let signature = decode_fixed(string_field(document, "signature").unwrap_or(""), 64)?;
    let public_key = decode_fixed(context.signer_public_key, 32)?;
    let mut unsigned = document.clone();
    unsigned.remove("signature");
    let mut message = SIGNATURE_DOMAIN.to_vec();
    message.extend(
        canonical_json_bytes(&Value::Object(unsigned))
            .map_err(|_| MutationRequestError::new("MUTATION_REQUEST_INVALID"))?,
    );
    let verifying_key = VerifyingKey::from_bytes(
        public_key
            .as_slice()
            .try_into()
            .map_err(|_| MutationRequestError::new("MUTATION_REQUEST_SIGNATURE_INVALID"))?,
    )
    .map_err(|_| MutationRequestError::new("MUTATION_REQUEST_SIGNATURE_INVALID"))?;
    let signature = Signature::from_bytes(
        signature
            .as_slice()
            .try_into()
            .map_err(|_| MutationRequestError::new("MUTATION_REQUEST_SIGNATURE_INVALID"))?,
    );
    verifying_key
        .verify_strict(&message, &signature)
        .map_err(|_| MutationRequestError::new("MUTATION_REQUEST_SIGNATURE_INVALID"))?;
    Ok(())
}

fn mutation_digest(document: &Map<String, Value>) -> Result<String, MutationRequestError> {
    let mut unsigned = document.clone();
    unsigned.remove("signature");
    unsigned.remove("digest");
    typed_digest(&Value::Object(unsigned))
}

fn typed_digest(value: &Value) -> Result<String, MutationRequestError> {
    Ok(format!(
        "sha256:{}",
        sha256_hex(
            &canonical_json_bytes(value)
                .map_err(|_| MutationRequestError::new("MUTATION_REQUEST_INVALID"))?
        )
    ))
}

fn reject_secrets(value: &Value) -> Result<(), MutationRequestError> {
    match value {
        Value::Object(map) => {
            for (key, nested) in map {
                let normalized: String = key
                    .chars()
                    .filter(|character| character.is_ascii_alphanumeric())
                    .map(|character| character.to_ascii_lowercase())
                    .collect();
                if !matches!(
                    normalized.as_str(),
                    "fencingtoken" | "signature" | "signaturealgorithm" | "signerkeyid"
                ) && [
                    "password",
                    "passwd",
                    "privatekey",
                    "ageidentity",
                    "apikey",
                    "accesskey",
                    "secretkey",
                    "token",
                    "credential",
                    "oauth",
                    "bearer",
                    "secret",
                ]
                .iter()
                .any(|fragment| normalized.contains(fragment))
                {
                    return Err(MutationRequestError::new(
                        "MUTATION_REQUEST_SECRET_DETECTED",
                    ));
                }
                reject_secrets(nested)?;
            }
        }
        Value::Array(items) => {
            for nested in items {
                reject_secrets(nested)?;
            }
        }
        Value::String(text) => {
            let lower = text.to_ascii_lowercase();
            if lower.contains("age-secret-key-")
                || lower.contains("-----begin") && lower.contains("private key")
            {
                return Err(MutationRequestError::new(
                    "MUTATION_REQUEST_SECRET_DETECTED",
                ));
            }
        }
        _ => {}
    }
    Ok(())
}

fn string_field<'a>(document: &'a Map<String, Value>, key: &str) -> Option<&'a str> {
    document.get(key).and_then(Value::as_str)
}

fn integer_field(document: &Map<String, Value>, key: &str) -> Result<i64, MutationRequestError> {
    document
        .get(key)
        .and_then(Value::as_i64)
        .ok_or_else(|| MutationRequestError::new("MUTATION_REQUEST_INVALID"))
}

fn decode_fixed(value: &str, size: usize) -> Result<Vec<u8>, MutationRequestError> {
    let raw = URL_SAFE_NO_PAD
        .decode(value)
        .map_err(|_| MutationRequestError::new("MUTATION_REQUEST_SIGNATURE_INVALID"))?;
    if raw.len() != size {
        return Err(MutationRequestError::new(
            "MUTATION_REQUEST_SIGNATURE_INVALID",
        ));
    }
    Ok(raw)
}

fn sha256_hex(value: &[u8]) -> String {
    let mut out = String::with_capacity(64);
    for byte in Sha256::digest(value) {
        use std::fmt::Write;
        let _ = write!(out, "{byte:02x}");
    }
    out
}

fn parse_utc_z(value: &str) -> Result<i64, MutationRequestError> {
    crate::authority_request::parse_utc_z_str(value)
        .map_err(|_| MutationRequestError::new("MUTATION_REQUEST_INVALID"))
}

fn valid_fleet_id(value: &str) -> bool {
    crate::authority_request::valid_fleet_id_str(value)
}

fn valid_hex64(value: &str) -> bool {
    value.len() == 64
        && value
            .bytes()
            .all(|byte| byte.is_ascii_hexdigit() && !byte.is_ascii_uppercase())
}

fn valid_control_id(value: &str) -> bool {
    let bytes = value.as_bytes();
    !bytes.is_empty()
        && bytes.len() <= 128
        && bytes[0].is_ascii_alphanumeric()
        && bytes[1..]
            .iter()
            .all(|byte| byte.is_ascii_alphanumeric() || matches!(byte, b'.' | b'_' | b':' | b'-'))
}

fn valid_signer_key_id(value: &str) -> bool {
    value.starts_with("ctrl-signer-")
        && value.len() == 28
        && value.as_bytes()[12..]
            .iter()
            .all(|byte| byte.is_ascii_hexdigit() && !byte.is_ascii_uppercase())
}
